src/ast/SemanticChecksVisitor.java

```diff
@@ -290,115 +309,215 @@ else if(ownerExprType.equals("ast.IdentifierExpr")){
+            if (forest.nameToClassDecl(((RefType) varType).id()) == null) visitResult = false;
         }
         //else than the type is not a reference so it's a legal type
     }
 
     @Override
     public void visit(BlockStatement blockStatement) {
-        for(Statement statement : blockStatement.statements()){
+        for (Statement statement : blockStatement.statements()) {
             statement.accept(this);
-            if(!visitResult) return;
+            if (!visitResult) return;
         }
     }
 
     @Override
     public void visit(IfStatement ifStatement) {
+        // req 16 - cond is boolean
+        AstType condType = getExprType(ifStatement.cond());
+        if (!(condType instanceof BoolAstType)) {
+            visitResult = false;
+            return;
+        }
+        Set<String> base = new HashSet<>(definitelyInitialized.peek());
+
+        definitelyInitialized.push(new HashSet<>(base));
+        ifStatement.thencase().accept(this);
+        definitelyInitialized.push(new HashSet<>(base));
+        ifStatement.elsecase().accept(this);
 
+        Set <String> branch2 =  definitelyInitialized.pop();
+        Set <String> branch1 =  definitelyInitialized.pop();
+        branch1.retainAll(branch2); // set intersection
+
+        // add the intersection back to the current scope
+        definitelyInitialized.peek().addAll(branch1);
     }
 
     @Override
     public void visit(WhileStatement whileStatement) {
-
+        // req 16 - cond is boolean
+        AstType condType = getExprType(whileStatement.cond());
+        if (!(condType instanceof BoolAstType)) {
+            visitResult = false;
+            return;
+        }
+        definitelyInitialized.push(new HashSet<>(definitelyInitialized.peek()));
+        whileStatement.body().accept(this);
+        // we don't know if the condition so we assume it didn't change anything in upper scope
+        definitelyInitialized.pop();
     }
 
     @Override
     public void visit(SysoutStatement sysoutStatement) {
-
+        sysoutStatement.arg().accept(this);
     }
 
     @Override
     public void visit(AssignStatement assignStatement) {
-
+        SymbolTable declTable = STLookup.findDeclTable(assignStatement.lv(), forest, assignStatement.enclosingScope(), programST);
+        VariableIntroduction assigneeDecl = (VariableIntroduction) STLookup.getDeclNode(declTable, assignStatement.lv());
+        AstType assigneeType = assigneeDecl.type();
+        AstType assignedValueType = getExprType(assignStatement.rv());
+        // TODO continue: req 15
+        assignStatement.rv().accept(this);
+        definitelyInitialized.peek().add(assignStatement.lv());
     }
 
     @Override
     public void visit(AssignArrayStatement assignArrayStatement) {
+        if (!definitelyInitialized.peek().contains(assignArrayStatement.lv())) {
+            visitResult = false;
+            return;
+        }
 
+        assignArrayStatement.index().accept(this);
```

Comment:
missing check - check to see int (req 21)
I should finish the adjustment to the expr to type method - will do so after this merge